algebird-core/src/main/scala/com/twitter/algebird/maps/redblack.scala

```diff
@@ -0,0 +1,219 @@
+      }
+      case (BNode(x, a, b), BNode(y, c, d)) => append(b, c) match {
+        case RNode(z, bb, cc) => rNode(z, bNode(x, a, bb), bNode(y, cc, d))
+        case bc => balanceLeft(x, a, bNode(y, bc, d))
+      }
+      case (a, RNode(x, b, c)) => rNode(x, append(a, b), c)
+      case (RNode(x, a, b), c) => rNode(x, a, append(b, c))
+    }
+
+    // NOTE: the balancing rules for node deletion all assume that the case of deleting a key
+    // that is not in the map is addressed elsewhere.  If these balancing functions are applied
+    // to a key that isn't present, they will fail destructively and uninformatively.
+    final protected def delLeft(node: INode[K], k: K) = node.lsub match {
+      case n: INode[K] if (n.color == B) => balanceLeft(node.data, node.lsub.del(k), node.rsub)
+      case _ => rNode(node.data, node.lsub.del(k), node.rsub)
+    }
+
+    final protected def delRight(node: INode[K], k: K) = node.rsub match {
+      case n: INode[K] if (n.color == B) => balanceRight(node.data, node.lsub, node.rsub.del(k))
+      case _ => rNode(node.data, node.lsub, node.rsub.del(k))
+    }
+  }
+
+  /** Represents a leaf node in the Red Black tree system */
+  trait LNode[K] extends Node[K] {
+    final def node(k: K) = None
+
+    final def nodeMin = None
+    final def nodeMax = None
+
+    final def ins(d: Data[K]) = rNode(d, this, this)
+    final def del(k: K) = this
+  }
+
+  private object LNode {
+    def unapply[K](node: LNode[K]): Boolean = true
+  }
+
+  /** Represents an internal node (Red or Black) in the Red Black tree system */
+  trait INode[K] extends Node[K] {
+    /** The Red/Black color of this node */
+    val color: Color
+    /** Including, but not limited to, the key */
+    val data: Data[K]
+    /** The left sub-tree */
+    val lsub: Node[K]
+    /** The right sub-tree */
+    val rsub: Node[K]
+
+    final def node(k: K) =
+      if (keyOrdering.lt(k, data.key)) lsub.node(k)
+      else if (keyOrdering.gt(k, data.key)) rsub.node(k)
+      else Some(this)
+
+    final def nodeMin = lsub match {
+      case LNode() => Some(this)
+      case _ => lsub.nodeMin
+    }
+
+    final def nodeMax = rsub match {
+      case LNode() => Some(this)
+      case _ => rsub.nodeMax
+    }
+
+    final def del(k: K) =
+      if (keyOrdering.lt(k, data.key)) delLeft(this, k)
+      else if (keyOrdering.gt(k, data.key)) delRight(this, k)
+      else append(lsub, rsub)
+
+    final def ins(d: Data[K]) =
+      if (color == R) {
+        if (keyOrdering.lt(d.key, data.key)) rNode(data, lsub.ins(d), rsub)
+        else if (keyOrdering.gt(d.key, data.key)) rNode(data, lsub, rsub.ins(d))
+        else rNode(d, lsub, rsub)
+      } else {
+        if (keyOrdering.lt(d.key, data.key)) balance(bNode(data, lsub.ins(d), rsub))
+        else if (keyOrdering.gt(d.key, data.key)) balance(bNode(data, lsub, rsub.ins(d)))
+        else bNode(d, lsub, rsub)
+      }
+  }
+
+  private object RNode {
```

Comment:
Actually, why not make:
```scala
case class RedNode[K](data: Data[K], lsub: INode[K], rsub: INode[K]) extends INode[K]
case class ...
```
or even:
```scala
case class INode[+C <: Color, K](color: C, lsub: INode[Color, K], rsub: INode[Color, K])
```
This would be more idiomatic.